db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectivityUtil.java

```diff
@@ -84,7 +84,10 @@ else if (outerCost.isAntiJoin())
 
 
     private static boolean isTheRightJoinPredicate(Predicate p, JoinPredicateType predicateType) {
-        if (p == null || !p.isJoinPredicate())
+        if (p == null || predicateType != JoinPredicateType.ALL && !p.isHashableJoinPredicate())
+            return false;
+
+        if (p.referencedSet.cardinality() < 2)
```

Comment:
Is it possible that a predicate has `predicateType == JoinPredicateType.ALL` but its `referencedSet.cardinality() < 2` ? I mean, if it references only one table or even no table, probably it shouldn't be `JoinPredicateType.ALL` at the first place. I'd like to understand in which situation we have such cases.

Reply:
JoinPredicateType.ALL isn't an attribute of the predicate, but the type of predicates we're considering.  This rule is already built into Predicate.isHashableJoinPredicate:

        return (leftUsableAsJoinColumn &&
                rightUsableAsJoinColumn &&
                leftTableNumber!=rightTableNumber) &&        <<-----------
                (!checkScope  || !isScopedForPush());

So, now we are considering more than hashable predicates, but we still need to enforce the rule that we are only looking for predicates referencing more than one table.